File: src/main/java/leetcode/_200/SolutionUF.java
package leetcode._200;

// 4ms
public class SolutionUF {
    public int numIslands(char[][] grid) {
        return new UnionFind(grid).count();
    }
}

class UnionFind {

    private final int[] id;
    private final int[] rank;
    private int count;

    public UnionFind(char[][] grid) {
        var N = grid.length;
        var M = grid[0].length;

        id = new int[N * M];
        rank = new int[N * M];

        for (var i = 0; i < N; i++) {
            for (var j = 0; j < M; j++) {
                if (grid[i][j] == '1') {
                    var p = i * M + j;
                    id[p] = p;
                    count++;

                    // try to connect prev component on x axe
                    var x = i - 1;
                    if (x >= 0 && grid[x][j] == '1') {
                        var q = x * M + j;
                        union(p, q);
                    }

                    // try to connect prev component on y axe
                    var y = j - 1;
                    if (y >= 0 && grid[i][y] == '1') {
                        var q = i * M + y;
                        union(p, q);
                    }
                }
            }
        }
    }

    // merges two components if the two sites are in different components,
    // using rank to rename component
    public void union(int p, int q) {
        int pID = find(p);
        int qID = find(q);

        // Nothing to do if p and q are already
        // in the same component.
        if (pID == qID)
            return;

        // Rename component to the highest rank's name between components.
        if (rank[pID] > rank[qID]) {
            id[qID] = pID;
        } else if (rank[pID] < rank[qID]) {
            id[pID] = qID;
        } else {
            id[qID] = pID;
            rank[pID] += 1;
        }
        --count;
    }

    public int find(int p) {
        while (p != id[p])
            p = id[p];
        return p;
    }

    // method returns the number of components
    public int count() {
        return count;
    }

}

File: src/main/java/leetcode/_200/solution_uf.go
package main

import "fmt"

func numIslands(grid [][]byte) int {
	N := len(grid)
	M := len(grid[0])

	uf := NewUF(M, N)

	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			if grid[i][j] == '1' {
				p := i*M + j
				uf.id[p] = p
				uf.count++

				// try to connect prev component on x axe
				x := i - 1
				if x >= 0 && grid[x][j] == '1' {
					q := x*M + j
					uf.Union(p, q)
				}

				// try to connect prev component on y axe
				y := j - 1
				if y >= 0 && grid[i][y] == '1' {
					q := i*M + y
					uf.Union(p, q)
				}
			}

		}
	}

	return uf.Count()
}

type UnionFind struct {
	id    []int
	rank  []int
	count int
}

func NewUF(m int, n int) UnionFind {
	return UnionFind{
		id:    make([]int, n*m),
		rank:  make([]int, n*m),
		count: 0,
	}
}

// find returns an integer component identifier for a given site
func (uf *UnionFind) find(p int) int {
	for p != uf.id[p] {
		p = uf.id[p]
	}
	return p
}

// Count returns the number of components
func (uf *UnionFind) Count() int {
	return uf.count
}

// Union merges two components if the two sites are in different components
func (uf *UnionFind) Union(p int, q int) {
	pId := uf.find(p)
	qId := uf.find(q)

	// Nothing to do if p and q are already in the same component.
	if pId == qId {
		return
	}

	// Rename component to the highest rank's name between components.
	if uf.rank[pId] > uf.rank[qId] {
		uf.id[qId] = pId
	} else if uf.rank[pId] < uf.rank[qId] {
		uf.id[pId] = qId
	} else {
		uf.id[qId] = pId
		uf.rank[pId] += 1
	}

	uf.count--
}

// ***************** Testing *****************
func main() {
	grid := [][]byte{
		{'1', '1', '0', '0', '0'},
		{'1', '1', '0', '0', '0'},
		{'0', '0', '1', '0', '0'},
		{'0', '0', '0', '1', '1'},
	}

	res := numIslands(grid)
	fmt.Println(res)
}
